[SQL] Rust Error when applying ABS(interval_value)

The following SQL:
```
CREATE TABLE tbl(tmestmp TIMESTAMP);

CREATE LOCAL VIEW intervalv AS SELECT
(tmestmp - TIMESTAMP '2025-06-21 14:23:44')YEAR AS yr
FROM tbl;

CREATE MATERIALIZED VIEW v AS SELECT
CAST(ABS(yr) AS BIGINT)
FROM intervalv;
```

Given following input: 
**"tmestmp": "2020-06-21 14:23:44"**

Produces the following Rust Error:
**error[E0425]: cannot find function `abs_LongInterval_YEARSN` in this scope
    --> crates/feldera_pipe_operator_06650a217d183c93/src/lib.rs:68:105
     |
68   | ...gInterval_YEARSN(abs_LongInterval_YEARSN([p1.0.as](http://p1.0.as/)_ref().cloned()))), 12i64));
     |                     ^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `cast_to_s_LongInterval_YEARSN`
     |
    ::: /Users/rivuadhikari/feldera/crates/sqllib/src/casts.rs:1553:1
     |
1553 | cast_to_string!(LongInterval_YEARS, LongInterval);
     | ------------------------------------------------- similarly named function `cast_to_s_LongInterval_YEARSN` defined here**

**Note:**
CAST(ABS(yr) AS DOUBLE) produces this Runtime error instead:

> _**RuntimeError:
> Pipeline test failed to compile:
> Error in SQL statement
> Cast function cannot convert value of type INTERVAL YEAR to type DOUBLE
> Code snippet:
>    49|CAST(ABS(yr) AS DOUBLE)
>       ^^^^^^^^^^^^^^^^^^^^^^^
>    50|FROM intervalv;**_

--- a/crates/sqllib/src/interval.rs
+++ b/crates/sqllib/src/interval.rs
@@ -75,6 +75,13 @@ impl ShortInterval {
 }
 
 #[doc(hidden)]
+pub fn abs_ShortInterval(value: ShortInterval) -> ShortInterval {
+    ShortInterval::new(num::abs(value.milliseconds))
+}
+
+some_polymorphic_function1!(abs, ShortInterval, ShortInterval, ShortInterval);
+
+#[doc(hidden)]
 /// This function is used in rolling window computations, which require all
 /// values to be expressed using unsigned types.
 pub fn to_bound_ShortInterval_Date_u128(value: &ShortInterval) -> u128 {
@@ -442,6 +449,13 @@ impl LongInterval {
         (months / 12) * mul
     }
 }
+
+#[doc(hidden)]
+pub fn abs_LongInterval(value: LongInterval) -> LongInterval {
+    LongInterval::new(num::abs(value.months))
+}
+
+some_polymorphic_function1!(abs, LongInterval, LongInterval, LongInterval);
 
 /// Multiply a `LongInterval` by an integer producing a `LongInterval`
 impl Mul<i32> for LongInterval {

--- a/docs/sidebars.js
+++ b/docs/sidebars.js
@@ -182,29 +182,35 @@ const sql = {
         'sql/grammar',
         'sql/identifiers',
         'sql/types',
-        'sql/casts',
-        'sql/operators',
-        'sql/aggregates',
-        'sql/boolean',
-        'sql/comparisons',
-        'sql/integer',
-        'sql/float',
-        'sql/decimal',
-        'sql/string',
-        'sql/uuid',
-        'sql/binary',
-        'sql/array',
-        'sql/map',
-        'sql/datetime',
-        'sql/json',
+        {
+            type: 'category',
+            label: 'operations',
+            items: [
+                'sql/aggregates',
+                'sql/array',
+                'sql/binary',
+                'sql/boolean',
+                'sql/casts',
+                'sql/comparisons',
+                'sql/datetime',
+                'sql/decimal',
+                'sql/float',
+                'sql/function-index'
+                'sql/integer',
+                'sql/json',
+                'sql/map',
+                'sql/operators',
+                'sql/string',
+                'sql/table',
+                'sql/uuid',
+            ]
+        },
         'sql/materialized',
         'sql/recursion',
         'sql/ad-hoc',
         'sql/streaming',
         'sql/system',
-        'sql/table',
-        'sql/udf',
-        'sql/function-index'
+        'sql/udf'
     ]
 };
 

--- a/docs/sql/array.md
+++ b/docs/sql/array.md
@@ -1,4 +1,4 @@
-# Operations on arrays
+# Array operations
 
 An array type can be created by applying the `ARRAY` suffix to another
 type.  For example `INT ARRAY` is an array of integers.  Array indexes

--- a/docs/sql/casts.md
+++ b/docs/sql/casts.md
@@ -1,4 +1,4 @@
-# Data Type Conversions
+# Casts and Data Type Conversions
 
 SQL expressions can mix data of different types in the same
 expression.  Most operations, however, require operands to have either

--- a/docs/sql/datetime.md
+++ b/docs/sql/datetime.md
@@ -352,10 +352,10 @@ Arithmetic between a `DATE` and an `INTERVAL` first converts the
 interval to a whole number days (rounding down) and then performs the
 computation on whole days.
 
-<a id="date_sub"></a>
-`DATE_SUB` is a synonim for `DATE` - `INTERVAL`.
-<a id="date_add"></a>
-`DATE_ADD` is a synonim for `DATE` + `INTERVAL`.
+<a id="date_sub"></a> `DATE_SUB` is a synonim for `DATE` - `INTERVAL`.
+<a id="date_add"></a> `DATE_ADD` is a synonim for `DATE` + `INTERVAL`.
+<a id="abs"></a>`ABS`(interval) computes the absolute value of an
+interval.
 
 ## Timezones
 

--- a/docs/sql/decimal.md
+++ b/docs/sql/decimal.md
@@ -1,4 +1,4 @@
-# Operations on Decimals
+# Decimal Operations
 
 A synonym for the ``decimal`` type is ``numeric``.
 

--- a/docs/sql/json.md
+++ b/docs/sql/json.md
@@ -1,4 +1,4 @@
-# Dynamically-Typed Values and JSON Support
+# JSON and Dynamically-Typed Value Support
 
 ## Direct conversion of JSON strings to User-Defined data types
 

--- a/docs/sql/map.md
+++ b/docs/sql/map.md
@@ -1,4 +1,4 @@
-# Operations on Maps
+# Map Operations
 
 A MAP type can be created using the syntax `MAP<type, type>`.
 For example `MAP<VARCHAR, INT>` is a map from strings to integers.

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -977,9 +977,35 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
 
                         return compilePolymorphicFunction(true, call, node, type, Linq.list(left, right), 2);
                     }
-                    case "numeric_inc":
-                    case "sign":
                     case "abs": {
+                        String opName1 = getCallName(call);
+                        validateArgCount(node, opName1, ops.size(), 1);
+                        StringBuilder functionName = new StringBuilder(opName1);
+                        DBSPExpression[] operands = ops.toArray(new DBSPExpression[0]);
+                        boolean resultNullable = false;
+                        for (DBSPExpression op: ops) {
+                            DBSPType type1 = op.getType();
+                            if (type1.is(IsIntervalType.class)) {
+                                var base = type1.to(DBSPTypeBaseType.class);
+                                functionName.append("_").append(base.shortName())
+                                        .append(base.nullableSuffix());
+                            } else {
+                                functionName.append("_").append(type1.baseTypeWithSuffix());
+                            }
+                            if (type1.mayBeNull)
+                                resultNullable = true;
+                        }
+                        DBSPType intermediateResultType = type;
+                        if (type.mayBeNull && !resultNullable) {
+                            intermediateResultType = type.withMayBeNull(false);
+                        }
+                        DBSPExpression result = new DBSPApplyExpression(node, functionName.toString(), intermediateResultType, operands);
+                        if (type.sameType(intermediateResultType))
+                            return result;
+                        return result.castToNullable();
+                    }
+                    case "numeric_inc":
+                    case "sign": {
                         return compilePolymorphicFunction(true, call, node, type,
                                 ops, 1);
                     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/CalciteFunctions.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/CalciteFunctions.java
@@ -169,7 +169,7 @@ public class CalciteFunctions implements FunctionDocumentation.FunctionRegistry 
             new Func(SqlStdOperatorTable.MOD, "MOD", SqlLibrary.STANDARD, "integer#mod", false),
             new Func(SqlStdOperatorTable.LN, "LN", SqlLibrary.STANDARD, "decimal#ln,float#ln", false),
             new Func(SqlStdOperatorTable.LOG10, "LOG10", SqlLibrary.STANDARD, "decimal#log10,float#log10", false),
-            new Func(SqlStdOperatorTable.ABS, "ABS", SqlLibrary.STANDARD, "decimal#abs,float#abs,integer#abs", false),
+            new Func(SqlStdOperatorTable.ABS, "ABS", SqlLibrary.STANDARD, "decimal#abs,float#abs,integer#abs,datetime#abs", false),
 
             new Func(SqlStdOperatorTable.ACOS, "ACOS", SqlLibrary.STANDARD, "float#acos", false),
             new Func(SqlStdOperatorTable.ASIN, "ASIN", SqlLibrary.STANDARD, "float#asin", false),

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
@@ -1,0 +1,29 @@
+package org.dbsp.sqlCompiler.compiler.sql.simple;
+
+import org.dbsp.sqlCompiler.compiler.sql.tools.SqlIoTest;
+import org.junit.Test;
+
+public class Regression1Tests extends SqlIoTest {
+    @Test
+    public void issue3952() {
+        var ccs = this.getCCS("""
+                CREATE TABLE tbl(t0 TIMESTAMP, t1 TIMESTAMP NOT NULL);
+                
+                CREATE LOCAL VIEW intervalv AS SELECT
+                (t0 - TIMESTAMP '2025-06-21 14:23:44') YEAR AS yr0,
+                (TIMESTAMP '2025-06-21 14:23:44' - t1) YEAR AS yr1,
+                (t0 - TIMESTAMP '2025-06-21 14:23:44') DAYS AS d0,
+                (TIMESTAMP '2025-06-21 14:23:44' - t1) DAYS AS d1
+                FROM tbl;
+                
+                CREATE MATERIALIZED VIEW v AS SELECT
+                CAST(ABS(yr0) AS BIGINT), CAST(ABS(yr1) AS BIGINT),
+                CAST(ABS(d0) AS BIGINT), CAST(ABS(d1) AS BIGINT)
+                FROM intervalv;""");
+        ccs.step("INSERT INTO tbl VALUES(TIMESTAMP '2020-01-01 00:00:00', TIMESTAMP '2019-01-01 00:00:00')",
+                """
+                  v0 | v1 |  v2  |  v3  | weight
+                 ----------------------------
+                   5 |  6 | 1998 | 2363 | 1""");
+    }
+}

